creator/ingest_runs/genomic_data_loader.py

```diff
@@ -0,0 +1,356 @@
+        self.logger.info(
+            f"Loading {df.shape[0]} harmonized genomic files into "
+            f"{DATASERVICE_URL}"
+        )
+
+        # Load harmonized genomic files
+        genomic_cache = self._load_entities(GEN_FILE, df)
+        genomic_df = self._get_genomic_file_kf_ids(genomic_cache, genomic_df)
+
+        return genomic_df
+
+    def load_specimen_harmonized_gf_links(self, genomic_df):
+        """
+        Step 2
+
+        Link the harmonized genomic files to their related specimens by loading
+        the appropriate biospecimen-genomic-file records into the Data Service
+        """
+        # Prepare data
+        biospec_gen_df = genomic_df[[
+            CONCEPT.GENOMIC_FILE.ID,
+            CONCEPT.BIOSPECIMEN.TARGET_SERVICE_ID,
+            CONCEPT.GENOMIC_FILE.TARGET_SERVICE_ID,
+        ]]
+        biospec_gen_df[CONCEPT.BIOSPECIMEN.ID] = biospec_gen_df[
+            CONCEPT.BIOSPECIMEN.TARGET_SERVICE_ID
+        ]
+        biospec_gen_df[CONCEPT.BIOSPECIMEN_GENOMIC_FILE.VISIBLE] = True
+        df = biospec_gen_df.drop_duplicates(
+            [CONCEPT.BIOSPECIMEN.TARGET_SERVICE_ID,
+             CONCEPT.GENOMIC_FILE.TARGET_SERVICE_ID],
+        )
+
+        # Load into Data Service
+        self.logger.info(
+            f"Creating {df.shape[0]} biospecimen-genomic-file links in "
+            f"{DATASERVICE_URL}"
+        )
+        _ = self._load_entities(BIO_GEN_FILE, df)
+
+        return biospec_gen_df
+
+    def load_seq_exp_harmonized_genomic_files(self, genomic_df):
+        """
+        Step 3
+
+        Link the harmonized genomic files to their related sequencing
+        experiments by loading the appropriate
+        sequencing-experiment-genomic-file records into the Data Service
+
+        This will have to be done by getting the sequencing experiments that
+        are linked to the source genomic files used to produce the harmonized
+        genomic files.
+        """
+        # Get sequencing experiments for the unharmonized genomic files in
+        # the study
+        se_source_gf_df = self._get_seq_experiment_genomic_files(
+            harmonized=False
+        )
+        # Merge ^ with the genomic df which as both harmonized and unharmonized
+        # gfs in one table
+        se_gf_df = pd.merge(
+            genomic_df, se_source_gf_df, on=CONCEPT.GENOMIC_FILE.SOURCE_FILE
+        )
+        se_gf_df[CONCEPT.SEQUENCING_GENOMIC_FILE.VISIBLE] = True
+
+        # Load the harmonized genomic-file sequencing-experiment links
+        df = se_gf_df.drop_duplicates(
+            [CONCEPT.SEQUENCING.TARGET_SERVICE_ID,
+             CONCEPT.GENOMIC_FILE.TARGET_SERVICE_ID],
+        )
+        self.logger.info(
+            f"Creating {df.shape[0]} sequencing-experiment-genomic-file "
+            f"links in {DATASERVICE_URL}"
+        )
+        _ = self._load_entities(SEQ_EXP_GEN_FILE, df)
+
+        return se_gf_df
+
+    def _prep_harmonized_gf_df(self, df):
```

Comment:
This is a totally valid point. The file we used to ingest the unharmonized genomic files in the fake test study has a different format from the genomic workflow output manifests, so the same extract config can't be used again, but it should be possible to just write an extract config for this initial step of the process.

Reply:
We might still be able to use an extract config to perform these transformations by making use of the ingest lib's ExtractStage here. Idk how hard or easy that will be / if changes are be needed in the ingest library since the ExtractStage wasn't originally designed to be a standalone utility.

Regardless of how the transformations happen here, I don't think it will affect our use of the ingest cache? Unless I'm missing something